### PR TITLE
Refactor authentication

### PR DESCRIPTION
**Pull request by fremartini**

### CODE REVIEW

**Review by TTA777 — CHANGES_REQUESTED** [review]
Minor stuff, but not a fan of the nullable user

File: lib/features/authentication/data/models/authenticated_user_model.dart [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import 'package:coffeecard/features/authentication/domain/entities/authenticated_user.dart';
+
+class AuthenticatedUserModel extends AuthenticatedUser {
+  const AuthenticatedUserModel({
+    required super.email,
+    required super.token,
+    required super.encodedPasscode,
+  });
+
+  factory AuthenticatedUserModel.fromJson(Map<String, dynamic> json) {
```
[the remark this answers is not in the document]
Comment (fremartini) [reply lib/features/authentication/data/models/authenticated_user_model.dart:10]: Well Dart is a bit special when it comes to serialization and deserialization. The simplest solution is to use handwritten constructors for parsing raw json as done here. Another approach is to use autogenerated code such as with https://pub.dev/packages/json_serializable but that is beyond the scope here,  we might consider doing this at another time.  When you use json.decode as we do in this project, Dart looks if the function implements a toJson() function and uses that if it does so we can parse it like done here. So this solution should suffice for now

File: lib/features/authentication/data/datasources/authentication_local_data_source.dart [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  ) async {
+    await storage.write(
+      key: _authenticatedUserKey,
+      value: json.encode(authenticatedUser),
+    );
+
+    logger.d('$authenticatedUser added to storage');
+  }
+
+  Future<AuthenticatedUserModel?> getAuthenticatedUser() async {
```
[the remark this answers is not in the document]
Comment (fremartini) [reply lib/features/authentication/data/datasources/authentication_local_data_source.dart:29]: In this case we need to differentiate between a user not being there (if they logged out or never have been logged in), and an error having occured. So we might get nothing when reading from storage but that shouldnt indicate an error

[the remark this answers is not in the document]
Comment (TTA777) [reply lib/features/authentication/data/datasources/authentication_local_data_source.dart:29]: How about an optional, then? Doesn't have to be an either, I just don't like null when I can avoid it 😛

**Review by TTA777 — APPROVED** [review]
lgtm

**Pull request merged**